### PR TITLE
MGTransferBlockMatrixFree for LA::d::BlockVector (version 2)

### LINKED ISSUE #4574: equivalent of MGTransferMatrixFree for LinearAlgebra::distributed::BlockVector
I plan to add `MGTransferMatrixFree` equivalent for`LinearAlgebra::distributed::BlockVector` so that transfer is done for each block simultaneously. I am interested in the case where each block lives in the whole FE space. In other words, the result of the transfer should be the same as doing current `MGTransferMatrixFree` for each element in `std::vector< LinearAlgebra::distributed::Vector>` separately. 

The question is, if anyone is already working on those things or have such functionality implemented? 
If not, i will start digging into `MGTransferMatrixFree` and its parent classes to figure out how to do it.
Essentially, i expect it to be a matter of wrapping current transfer operations into `for (b = 0; b < v.n_blocks(); ++b)`.

@guidokanschat @kronbichler @masterleinad ping.

### PR DESCRIPTION
**Pull request by davydden**
alternative to https://github.com/dealii/dealii/pull/4667

fixes #4574

### CODE REVIEW

File: source/multigrid/mg_transfer_matrix_free.cc [hunk 1]
```diff
@@ -553,6 +553,96 @@ MGTransferMatrixFree<dim,Number>::memory_consumption() const
...
+  matrix_free_transfer.clear();
+}
+
+
+
+template <int dim, typename Number>
+void MGTransferBlockMatrixFree<dim,Number>::build
+(const DoFHandler<dim,dim>  &mg_dof)
+{
+  this->fill_and_communicate_copy_indices(mg_dof);
```
Comment (davydden) [inline source/multigrid/mg_transfer_matrix_free.cc]: i must admit I am very unsure about calling parent's methods.

**Comment by kronbichler** [comment]
I was originally thinking about a different design: I thought of creating `MGTransferBlockMatrixFree` that has only the basic interface to transfer classes but does not have the same inheritance structure of the other transfer classes, but simply holds a `MGTransferMatrixFree` as a member variable and calls that classes' methods. That would avoid duplicating some code and adding the additional `MGLevelGlobalTransfer`. Have you thought about that? Apart from that, I think that it makes sense to distinguish the classes by the names. This is also what we do for the matrices and some other transfers, even though the address other situations.

**Comment by davydden** [comment]
Oh, I see. Yes, that makes perfect sense. I will update the PR.

**Comment by davydden** [comment]
@kronbichler i updated and added a unit test. I think that's what you suggested.

**Comment by kronbichler** [comment]
Very nice, this looks like what I was thinking of indeed, yes. Before we go to the details, I think you want to also add the `copy_to_mg` and `copy_from_mg` methods, right? I would appreciate if we could have a test that runs a full MG problem, similar to what `matrix_free/parallel_multigrid_adaptive_01` (or something similar) is doing. One could set up a Laplace operator that is based on a scalar `DoFHandler` but uses two components inside a `BlockVector` which are solved within a single solver.

**Comment by davydden** [comment]
@kronbichler i added copy_to/from_mg. 

> One could set up a Laplace operator that is based on a scalar DoFHandler but uses two components inside a BlockVector which are solved within a single solver.

can you please expand on this one? I have a feeling that one would need to add similar wrappers for smoothers as well as matrix-free operator. Or is there a way to do what you suggest differently?

**Comment by davydden** [comment]
@kronbichler i added the GMG test and it seems to be working as expected (see tests). So this PR should be good for another round of review.

File: tests/matrix_free/parallel_multigrid_adaptive_06.with_mpi=true.with_p4est=true.mpirun=7.output [hunk 2]
```diff
@@ -0,0 +1,97 @@
+
```
Comment (kronbichler) [inline tests/matrix_free/parallel_multigrid_adaptive_06.with_mpi=true.with_p4est=true.mpirun=7.output]: Do we really need two variants of this test? I think it would be enough to have the test output with 3 ranks and delete this one. The reason we had 7 ranks in the original test is that there were some corner cases where the transfer missed some weights in the preliminary implementation, but here we simply want to see that the block vector gets the right partitions etc which we should also have with 3 ranks, shouldn't we?

File: tests/matrix_free/parallel_multigrid_adaptive_06ref.with_mpi=true.with_p4est=true.mpirun=7.output [hunk 3]
```diff
@@ -0,0 +1,49 @@
+
```
Comment (kronbichler) [inline tests/matrix_free/parallel_multigrid_adaptive_06ref.with_mpi=true.with_p4est=true.mpirun=7.output]: same here

File: tests/multigrid/transfer_05.with_mpi=true.with_p4est=true.mpirun=3.output [hunk 4]
```diff
@@ -0,0 +1,367 @@
+
```
Comment (kronbichler) [inline tests/multigrid/transfer_05.with_mpi=true.with_p4est=true.mpirun=3.output]: Do we really need two variants of this test? I think it would be enough to have this test output and delete the one with `mpirun=1`.

File: tests/multigrid/transfer_matrix_free_02_block.with_mpi=true.with_p4est=true.with_trilinos=true.mpirun=5.output [hunk 5]
```diff
@@ -0,0 +1,259 @@
+
```
Comment (kronbichler) [inline tests/multigrid/transfer_matrix_free_02_block.with_mpi=true.with_p4est=true.with_trilinos=true.mpirun=5.output]: Same here: I would delete this one and only keep the single processor case (or the other way around).

**Comment by davydden** [comment]
@kronbichler i agree with removing additional test output. The PR's been updated.

**Comment by kronbichler** [comment]
Thanks for the adjustments. :+1:
